File: scripts/scrape-chrono24-dealer.py
#!/usr/bin/env python3
"""
Chrono24 Dealer Scraper — powered by Scrapling (replaces FlareSolverr)
Usage: python3 scripts/scrape-chrono24-dealer.py <dealer-slug>
Example: python3 scripts/scrape-chrono24-dealer.py rccrown

Scrapling's StealthyFetcher bypasses Cloudflare automatically.
No Docker, no FlareSolverr needed.
"""

import sys
import re
import os
import time
from pathlib import Path
from dotenv import load_dotenv

# Load .env.local
load_dotenv(Path(__file__).parent.parent / ".env.local")

from scrapling.fetchers import StealthyFetcher
from supabase import create_client

# ── Config ────────────────────────────────────────────────────────────────────
SUPABASE_URL = os.environ["NEXT_PUBLIC_SUPABASE_URL"]
SUPABASE_KEY = os.environ["SUPABASE_SERVICE_ROLE_KEY"]
PAGE_SIZE = 60

DEALER_BLOCKLIST = ["jewelsintimeofboca"]

ALLOWED_BRANDS = [
    "rolex", "richard mille", "patek philippe", "patek",
    "vacheron constantin", "vacheron", "f.p. journe", "fp journe",
    "audemars piguet", "audemars",
]

sb = create_client(SUPABASE_URL, SUPABASE_KEY)


# ── Helpers ───────────────────────────────────────────────────────────────────

def is_allowed_brand(title: str) -> bool:
    if not title:
        return False
    t = title.lower()
    return any(t.startswith(b) for b in ALLOWED_BRANDS)


def parse_price(text: str) -> int | None:
    match = re.search(r"\$([\d,]+)", text)
    return int(match.group(1).replace(",", "")) if match else None


def fetch_page(url: str) -> object:
    """Fetch a page using Scrapling's StealthyFetcher (bypasses Cloudflare)."""
    print(f"  → Fetching: {url}")
    page = StealthyFetcher.fetch(
        url,
        headless=True,
        network_idle=True,
        timeout=45000,
        block_images=True,   # faster — we don't need images rendered
        block_webfonts=True,
    )
    return page


def get_merchant_id(slug: str) -> int:
    """Get merchant ID from Chrono24. Works with headless=True (no FlareSolverr needed)."""
    # Try search page with dealer slug — more reliable than /dealer/ URL
    search_url = f"https://www.chrono24.com/search/index.htm?dosearch=true&dealer={slug}&p=1&pageSize=1"
    page = fetch_page(search_url)
    html = page.html_content

    for pattern in [
        r'"contactDealerLayerMerchantId":\s*(\d+)',
        r'merchantId["\s:=]+(\d{5,})',
        r'"merchant_id":\s*(\d+)',
        r'data-merchant-id=["\'](\d+)',
    ]:
        match = re.search(pattern, html, re.IGNORECASE)
        if match:
            return int(match.group(1))

    # Fallback: try the /dealer/ page
    page2 = fetch_page(f"https://www.chrono24.com/dealer/{slug}/index.htm")
    html2 = page2.html_content
    for pattern in [
        r'"contactDealerLayerMerchantId":\s*(\d+)',
        r'merchantId["\s:=]+(\d{5,})',
    ]:
        match = re.search(pattern, html2, re.IGNORECASE)
        if match:
            return int(match.group(1))

    raise ValueError(f"Could not find merchantId for dealer: {slug}")


def get_dealer_name(slug: str) -> str:
    try:
        page = fetch_page(f"https://www.chrono24.com/dealer/{slug}/index.htm")
        html = page.html_content

        # Try h1
        h1 = page.css("h1")
        if h1:
            return h1.first.text.strip()

        match = re.search(r'"dealerName":\s*"([^"]+)"', html)
        if match:
            return match.group(1).strip()
    except Exception:
        pass
    return slug


def parse_cards(page) -> list[dict]:
    """Parse listing cards from a Chrono24 search results page."""
    results = []

    # Primary selector — Chrono24 article cards
    cards = page.css('[class*="js-article-item-container"]')

    if not cards:
        # Fallback: try alternative selectors (Scrapling adaptive)
        cards = page.css('[data-article-id]') or page.css('.article-item')

    for card in cards:
        try:
            link = card.css('a[href*="--id"]')
            if not link:
                continue

            href = link.first.attrib.get("href", "")
            id_match = re.search(r"--id(\d+)\.htm", href)
            if not id_match:
                continue

            chrono24_id = id_match.group(1)

            # Title from img alt or link text
            img = card.css("img[alt]")
            title = ""
            img_url = ""
            if img:
                title = img.first.attrib.get("alt", "").strip()
                img_url = (
                    img.first.attrib.get("data-lazy-sweet-spot-master-src", "")
                    .replace("_SIZE_", "280")
                    or img.first.attrib.get("src", "")
                )

            if not title:
                title = link.first.text.strip()

            price = parse_price(card.text or "")
            url = href if href.startswith("http") else f"https://www.chrono24.com{href}"

            results.append({
                "chrono24Id": chrono24_id,
                "title": title,
                "price": price,
                "imgUrl": img_url,
                "url": url,
            })
        except Exception:
            continue

    return results


def scrape_all_pages(merchant_id: int) -> list[dict]:
    all_listings = []
    page_num = 1
    total_expected = None

    while True:
        url = (
            f"https://www.chrono24.com/search/index.htm"
            f"?dosearch=true&merchantId={merchant_id}"
            f"&p=1&pageSize={PAGE_SIZE}&showpage={page_num}&sortorder=1"
        )

        label = f"~{(total_expected // PAGE_SIZE) + 1}" if total_expected else "?"
        print(f"  Scraping page {page_num} of {label}...")

        page = fetch_page(url)

        # Parse total count on first page
        if page_num == 1:
            match = re.search(r'"numResult":\s*(\d+)', page.html_content)
            if match:
                total_expected = int(match.group(1))
                print(f"  Total listings found: {total_expected}")

        cards = parse_cards(page)
        print(f"  Page {page_num}: {len(cards)} cards parsed")

        if not cards:
            break

        all_listings.extend(cards)

        if total_expected and len(all_listings) >= total_expected:
            break
        if len(cards) < PAGE_SIZE:
            break

        page_num += 1
        time.sleep(2)  # respectful rate limit

    return all_listings


# ── Main ──────────────────────────────────────────────────────────────────────

def main():
    if len(sys.argv) < 2:
        print("Usage: python3 scripts/scrape-chrono24-dealer.py <dealer-slug>")
        sys.exit(1)

    slug = sys.argv[1].lower().strip()

    if slug in DEALER_BLOCKLIST:
        print(f"❌ Dealer '{slug}' is blocklisted — skipping.")
        sys.exit(0)

    print(f"\n🕷️  Scrapling — Chrono24 dealer: {slug}")
    print("   (No FlareSolverr needed — Scrapling handles Cloudflare bypass)")

    # Get merchant ID
    print("\nGetting merchant ID...")
    merchant_id = get_merchant_id(slug)
    print(f"Merchant ID: {merchant_id}")

    dealer_name = get_dealer_name(slug)
    print(f"Dealer name: {dealer_name}")

    # Upsert dealer record
    resp = sb.table("chrono24_dealers").upsert(
        {"merchant_id": merchant_id, "slug": slug, "name": dealer_name},
        on_conflict="merchant_id"
    ).execute()

    dealer_id = resp.data[0]["id"]
    print(f"Dealer DB ID: {dealer_id}")

    # Scrape all pages
    print("\nScraping inventory...")
    raw = scrape_all_pages(merchant_id)

    # Deduplicate
    seen = set()
    listings = []
    for l in raw:
        if l["chrono24Id"] not in seen:
            seen.add(l["chrono24Id"])
            listings.append(l)

    print(f"\nUnique listings: {len(listings)} ({len(raw) - len(listings)} duplicates removed)")

    # Detect sold listings
    existing = sb.table("chrono24_listings").select("chrono24_id").eq("merchant_id", merchant_id).eq("is_sold", False).execute()
    existing_ids = {r["chrono24_id"] for r in (existing.data or [])}
    scraped_ids = {l["chrono24Id"] for l in listings}
    sold_ids = list(existing_ids - scraped_ids)

    if sold_ids:
        print(f"\n🔴 Marking {len(sold_ids)} listings as sold...")
        from datetime import datetime, timezone
        sb.table("chrono24_listings").update({
            "is_sold": True,
            "sold_detected_at": datetime.now(timezone.utc).isoformat()
        }).in_("chrono24_id", sold_ids).execute()

    # Upsert current listings
    from datetime import datetime, timezone
    now = datetime.now(timezone.utc).isoformat()

    to_upsert = [
        {
            "chrono24_id": l["chrono24Id"],
            "dealer_id": dealer_id,
            "merchant_id": merchant_id,
            "title": (l["title"] or "")[:255],
            "price": l["price"],
            "currency": "USD",
            "image_url": (l["imgUrl"] or "")[:500] or None,
            "listing_url": (l["url"] or "")[:500] or None,
            "is_sold": False,
            "last_seen_at": now,
            "scraped_at": now,
        }
        for l in listings
        if l["chrono24Id"] and l["title"] and is_allowed_brand(l["title"])
    ]

    print(f"Upserting {len(to_upsert)} brand-filtered listings...")
    upserted = 0
    for i in range(0, len(to_upsert), 100):
        chunk = to_upsert[i:i+100]
        sb.table("chrono24_listings").upsert(chunk, on_conflict="chrono24_id").execute()
        upserted += len(chunk)
        print(f"  Upserted {upserted}/{len(to_upsert)}")

    # Update dealer metadata
    sb.table("chrono24_dealers").update({
        "total_listings": len(listings),
        "last_scraped_at": now,
        "updated_at": now,
    }).eq("id", dealer_id).execute()

    print(f"\n✅ Done!")
    print(f"   {len(listings)} listings scraped")
    print(f"   {len(to_upsert)} upserted (brand-filtered)")
    print(f"   {len(sold_ids)} marked as sold")


if __name__ == "__main__":
    main()

File: scripts/scrape-chrono24-dealer.mjs
// scripts/scrape-chrono24-dealer.mjs
// Usage: node scripts/scrape-chrono24-dealer.mjs <dealer-slug>
// Example: node scripts/scrape-chrono24-dealer.mjs jewelsintimeofboca
//
// FlareSolverr must be running at http://localhost:8191

import { createClient } from '@supabase/supabase-js';
import { config } from 'dotenv';
import { JSDOM } from 'jsdom';
config({ path: '.env.local' });

const sb = createClient(
  process.env.NEXT_PUBLIC_SUPABASE_URL,
  process.env.SUPABASE_SERVICE_ROLE_KEY
);

const FLARE_URL = 'http://localhost:8191/v1';
const PAGE_SIZE = 60;

// Only track these brands — all others are ignored
const ALLOWED_BRANDS = [
  'rolex',
  'richard mille',
  'patek philippe',
  'patek',
  'vacheron constantin',
  'vacheron',
  'f.p. journe',
  'fp journe',
  'audemars piguet',
  'audemars',
];

function isAllowedBrand(title) {
  if (!title) return false;
  const t = title.toLowerCase();
  return ALLOWED_BRANDS.some(b => t.startsWith(b));
}

async function flareGet(url) {
  const res = await fetch(FLARE_URL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ cmd: 'request.get', url, maxTimeout: 35000 }),
  });
  const data = await res.json();
  if (data.status !== 'ok') throw new Error(`FlareSolverr error: ${data.message}`);
  return data.solution.response;
}

function parsePrice(str) {
  const match = str.match(/\$([\d,]+)/);
  return match ? parseInt(match[1].replace(/,/g, ''), 10) : null;
}

/**
 * Parse all listing cards from a Chrono24 search results page.
 * Returns ALL cards (including cheap accessories) so pagination logic
 * can use raw card count to determine when to stop.
 * Price filtering happens later in the upsert pipeline.
 */
function parseCards(html) {
  const dom = new JSDOM(html);
  const doc = dom.window.document;
  const results = [];

  doc.querySelectorAll('[class*="js-article-item-container"]').forEach(card => {
    try {
      const link = card.querySelector('a[href*="--id"]');
      if (!link) return;
      const href = link.getAttribute('href');
      const idMatch = href.match(/--id(\d+)\.htm/);
      if (!idMatch) return;
      const chrono24Id = idMatch[1];

      const img = card.querySelector('img[alt]');
      const title = img?.getAttribute('alt') || card.querySelector('a[href*="--id"]')?.textContent?.trim() || '';
      const masterSrc = img?.getAttribute('data-lazy-sweet-spot-master-src') || '';
      const imgUrl = masterSrc
        ? masterSrc.replace('_SIZE_', '280')
        : (img?.getAttribute('src') || '');

      const price = parsePrice(card.textContent || '');
      const url = href.startsWith('/') ? `https://www.chrono24.com${href}` : href;

      // Include even if price is null/low — filter on upsert side
      results.push({ chrono24Id, title, price, imgUrl, url });
    } catch {
      // skip malformed card
    }
  });

  return results;
}

async function getMerchantId(slug) {
  console.log(`  Fetching dealer profile: https://www.chrono24.com/dealer/${slug}/index.htm`);
  const html = await flareGet(`https://www.chrono24.com/dealer/${slug}/index.htm`);
  const match = html.match(/"contactDealerLayerMerchantId":\s*(\d+)/);
  if (!match) throw new Error(`Could not find merchantId for dealer: ${slug}`);
  return parseInt(match[1], 10);
}

async function getDealerName(slug) {
  try {
    const html = await flareGet(`https://www.chrono24.com/dealer/${slug}/index.htm`);
    // Try H1 tag first
    const h1Match = html.match(/<h1[^>]*>([^<]+)<\/h1>/);
    if (h1Match) return h1Match[1].trim();
    // Fallback: look for "dealerName" in metaData JSON
    const nameMatch = html.match(/"dealerName":\s*"([^"]+)"/);
    if (nameMatch) return nameMatch[1].trim();
  } catch {
    // ignore
  }
  return slug;
}

async function scrapeAllPages(merchantId) {
  const allListings = [];
  let page = 1;
  let totalExpected = null;

  while (true) {
    // Chrono24 pagination: p=1 is fixed, showpage=N increments
    const url = `https://www.chrono24.com/search/index.htm?dosearch=true&merchantId=${merchantId}&p=1&pageSize=60&showpage=${page}&sortorder=1`;
    console.log(`  Scraping page ${page}${totalExpected ? ` of ~${Math.ceil(totalExpected / PAGE_SIZE)}` : ''}...`);

    const html = await flareGet(url);

    // Parse total count on first page
    if (page === 1) {
      const numMatch = html.match(/"numResult":\s*(\d+)/);
      totalExpected = numMatch ? parseInt(numMatch[1], 10) : null;
      if (totalExpected !== null) {
        console.log(`  Total listings found: ${totalExpected}`);
      }
    }

    const cards = parseCards(html);
    console.log(`  Page ${page}: ${cards.length} cards parsed (all price tiers)`);

    // Break only when no cards at all — NOT based on filtered count
    if (cards.length === 0) break;
    allListings.push(...cards);

    if (totalExpected !== null && allListings.length >= totalExpected) break;
    if (cards.length < PAGE_SIZE) break;

    page++;
    // Respectful rate limit between pages
    await new Promise(r => setTimeout(r, 2000));
  }

  return allListings;
}

async function main() {
  const slug = process.argv[2];
  if (!slug) {
    console.error('Usage: node scripts/scrape-chrono24-dealer.mjs <dealer-slug>');
    console.error('Example: node scripts/scrape-chrono24-dealer.mjs jewelsintimeofboca');
    process.exit(1);
  }

  console.log(`\n🔍 Scraping Chrono24 dealer: ${slug}`);

  // Get merchantId from dealer profile page
  console.log('Getting merchant ID...');
  const merchantId = await getMerchantId(slug);
  console.log(`Merchant ID: ${merchantId}`);

  // Get dealer name
  const dealerName = await getDealerName(slug);
  console.log(`Dealer name: ${dealerName}`);

  // Upsert dealer record
  const { data: dealer, error: dealerErr } = await sb
    .from('chrono24_dealers')
    .upsert(
      { merchant_id: merchantId, slug, name: dealerName },
      { onConflict: 'merchant_id' }
    )
    .select('id')
    .single();

  if (dealerErr) {
    console.error('Dealer upsert error:', dealerErr.message);
    process.exit(1);
  }
  console.log(`Dealer DB ID: ${dealer.id}`);

  // Scrape all pages
  console.log('\nScraping inventory...');
  const listings = await scrapeAllPages(merchantId);
  // Deduplicate by chrono24_id (pages can overlap)
  const seen = new Set();
  const deduped = listings.filter(l => {
    if (seen.has(l.chrono24Id)) return false;
    seen.add(l.chrono24Id);
    return true;
  });
  console.log(`\nTotal scraped: ${deduped.length} unique listings (${listings.length - deduped.length} duplicates removed)`);
  const listings_unique = deduped;

  // Get existing active listing IDs for sold detection
  const { data: existing } = await sb
    .from('chrono24_listings')
    .select('chrono24_id')
    .eq('merchant_id', merchantId)
    .eq('is_sold', false);

  const existingIds = new Set((existing ?? []).map(l => l.chrono24_id));
  const scrapedIds = new Set(listings_unique.map(l => l.chrono24Id));

  // Listings that disappeared = sold
  const soldIds = [...existingIds].filter(id => !scrapedIds.has(id));
  if (soldIds.length > 0) {
    console.log(`\n🔴 Marking ${soldIds.length} listings as sold...`);
    const { error: soldErr } = await sb
      .from('chrono24_listings')
      .update({ is_sold: true, sold_detected_at: new Date().toISOString() })
      .in('chrono24_id', soldIds);
    if (soldErr) console.error('Sold mark error:', soldErr.message);
  }

  // Upsert all current listings in batches of 100
  const now = new Date().toISOString();
  // Store all listings regardless of price — accessories included.
  // Consumers can filter by price threshold when querying.
  const toUpsert = listings_unique
    .filter(l => l.chrono24Id && l.title && isAllowedBrand(l.title)) // target brands only
    .map(l => ({
      chrono24_id: l.chrono24Id,
      dealer_id: dealer.id,
      merchant_id: merchantId,
      title: (l.title ?? '').substring(0, 255),
      price: l.price ?? null,
      currency: 'USD',
      image_url: (l.imgUrl ?? '').substring(0, 500) || null,
      listing_url: (l.url ?? '').substring(0, 500) || null,
      is_sold: false,
      last_seen_at: now,
      scraped_at: now,
    }));

  let upsertedCount = 0;
  for (let i = 0; i < toUpsert.length; i += 100) {
    const chunk = toUpsert.slice(i, i + 100);
    const { error: upsertErr } = await sb
      .from('chrono24_listings')
      .upsert(chunk, { onConflict: 'chrono24_id', ignoreDuplicates: false });
    if (upsertErr) {
      console.error(`Upsert error at chunk ${i}:`, upsertErr.message);
    } else {
      upsertedCount += chunk.length;
      console.log(`  Upserted ${upsertedCount}/${toUpsert.length}`);
    }
  }

  // Update dealer metadata
  await sb
    .from('chrono24_dealers')
    .update({
      total_listings: listings_unique.length,
      last_scraped_at: now,
      updated_at: now,
    })
    .eq('id', dealer.id);

  console.log(`\n✅ Done!`);
  console.log(`   ${listings_unique.length} listings saved`);
  console.log(`   ${soldIds.length} marked as sold`);
}

main().catch(err => {
  console.error('Fatal error:', err.message);
  process.exit(1);
});
